clippy_lints/src/format_args.rs

```diff
@@ -64,7 +66,66 @@ declare_clippy_lint! {
     "`to_string` applied to a type that implements `Display` in format args"
 }
 
-declare_lint_pass!(FormatArgs => [FORMAT_IN_FORMAT_ARGS, TO_STRING_IN_FORMAT_ARGS]);
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detect when a variable is not inlined in a format string,
+    /// and suggests to inline it.
+    ///
+    /// ### Why is this bad?
+    /// Non-inlined code is slightly more difficult to read and understand,
+    /// as it requires arguments to be matched against the format string.
+    /// The inlined syntax, where allowed, is simpler.
+    ///
+    /// ### Example
+    /// ```rust
+    /// # let var = 42;
+    /// # let width = 1;
+    /// # let prec = 2;
+    /// format!("{}", var);  // implied variables
+    /// format!("{0}", var);  // positional variables
+    /// format!("{v}", v=var);  // named variables
+    /// format!("{0} {0}", var);  // aliased variables
+    /// format!("{0:1$}", var, width);  // width support
+    /// format!("{0:.1$}", var, prec);  // precision support
+    /// format!("{:.*}", prec, var);  // asterisk support
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// # let var = 42;
+    /// # let width = 1;
+    /// # let prec = 2;
+    /// format!("{var}");  // implied, positional, and named variables
+    /// format!("{var} {var}");  // aliased variables
+    /// format!("{var:width$}");  // width support
+    /// format!("{var:.prec$}");  // precision and asterisk support
+    /// ```
+    ///
+    /// ### Known Problems
+    ///
+    /// Format string uses an indexed argument that cannot be inlined.
+    /// Supporting this case requires re-indexing of the format string.
+    ///
+    /// Until implemented, `print!("{0}={1}", var, 1+2)` should be changed to `print!("{var}={0}", 1+2)` by hand.
+    #[clippy::version = "1.64.0"]
+    pub NEEDLESS_FORMAT_ARGS,
+    nursery,
+    "using non-inlined variables in `format!` calls"
+}
+
+impl_lint_pass!(FormatArgs => [FORMAT_IN_FORMAT_ARGS, NEEDLESS_FORMAT_ARGS, TO_STRING_IN_FORMAT_ARGS]);
+
+pub struct FormatArgs {
+    format_args_capture: bool,
```

Comment:
```suggestion
    msrv: Option<RustcVersion>,
```

+ other changes as needed, moving the `meets_msrv(..)` to the `check_expr`